File: packages/sdk/src/resources/agent-services.ts
/**
 * Agent services resource client. Maps to `/v1/agent-services`.
 */
import type { AgentService } from "@settlekit/common";
import type { HttpClient, RequestOptions } from "../http-client.js";

/** Input for {@link AgentServicesResource.create}. */
export interface CreateAgentServiceInput {
  organizationId: string;
  merchantId: string;
  productId: string;
  name: string;
  description: string;
  endpoint: string;
  price: string;
  network?: "arc" | "base";
  inputSchema: Record<string, unknown>;
  outputSchema?: Record<string, unknown>;
}

/** Input for {@link AgentServicesResource.update}. */
export interface UpdateAgentServiceInput {
  name?: string;
  description?: string;
  endpoint?: string;
  price?: string;
}

/** Query options for {@link AgentServicesResource.list}. */
export interface ListAgentServicesParams {
  organizationId?: string;
}

/** Client for agent service endpoints. */
export class AgentServicesResource {
  constructor(private readonly http: HttpClient) {}

  /** Create an agent service listing. */
  create(input: CreateAgentServiceInput, options?: RequestOptions): Promise<AgentService> {
    return this.http.post<AgentService>("/v1/agent-services", input, options);
  }

  /** Discover agent services, optionally filtered by organization. */
  list(params: ListAgentServicesParams = {}, options?: RequestOptions): Promise<AgentService[]> {
    return this.http.get<AgentService[]>("/v1/agent-services", {
      ...options,
      query: { ...(params.organizationId !== undefined ? { organizationId: params.organizationId } : {}) },
    });
  }

  /** Retrieve an agent service by id. */
  retrieve(id: string, options?: RequestOptions): Promise<AgentService> {
    return this.http.get<AgentService>(`/v1/agent-services/${encodeURIComponent(id)}`, options);
  }

  /** Patch mutable agent service fields. */
  update(id: string, input: UpdateAgentServiceInput, options?: RequestOptions): Promise<AgentService> {
    return this.http.patch<AgentService>(`/v1/agent-services/${encodeURIComponent(id)}`, input, options);
  }

  /** Publish an agent service. */
  publish(id: string, options?: RequestOptions): Promise<AgentService> {
    return this.http.post<AgentService>(`/v1/agent-services/${encodeURIComponent(id)}/publish`, undefined, options);
  }

  /** Fetch the machine-readable metadata document (raw JSON, not enveloped). */
  metadata(id: string, options?: RequestOptions): Promise<Record<string, unknown>> {
    return this.http.get<Record<string, unknown>>(
      `/v1/agent-services/${encodeURIComponent(id)}/metadata.json`,
      options,
    );
  }
}

File: sdks/python/settlekit/resources/agent_services.py
"""Agent services resource — publish x402-priced AI services and fetch metadata."""

from __future__ import annotations

from typing import Any, Mapping

from ._base import AsyncResource, SyncResource


def _create_body(
    *,
    organization_id: str,
    merchant_id: str,
    product_id: str,
    name: str,
    description: str,
    endpoint: str,
    price: str,
    network: str,
    input_schema: Mapping[str, Any],
) -> dict[str, Any]:
    return {
        "organizationId": organization_id,
        "merchantId": merchant_id,
        "productId": product_id,
        "name": name,
        "description": description,
        "endpoint": endpoint,
        "price": price,
        "network": network,
        "inputSchema": dict(input_schema),
    }


class AgentServices(SyncResource):
    """Synchronous agent-services namespace."""

    def create(
        self,
        *,
        organization_id: str,
        merchant_id: str,
        product_id: str,
        name: str,
        description: str,
        endpoint: str,
        price: str,
        network: str,
        input_schema: Mapping[str, Any],
    ) -> Any:
        """Register an agent service. ``POST /v1/agent-services``."""
        return self._request(
            "POST",
            "/v1/agent-services",
            body=_create_body(
                organization_id=organization_id,
                merchant_id=merchant_id,
                product_id=product_id,
                name=name,
                description=description,
                endpoint=endpoint,
                price=price,
                network=network,
                input_schema=input_schema,
            ),
        )

    def list(self) -> Any:
        """List agent services. ``GET /v1/agent-services``."""
        return self._request("GET", "/v1/agent-services")

    def publish(self, service_id: str) -> Any:
        """Publish an agent service. ``POST /v1/agent-services/:id/publish``."""
        return self._request("POST", f"/v1/agent-services/{service_id}/publish")

    def metadata(self, service_id: str) -> Any:
        """Fetch machine-readable metadata. ``GET /v1/agent-services/:id/metadata.json``."""
        return self._request(
            "GET", f"/v1/agent-services/{service_id}/metadata.json"
        )


class AsyncAgentServices(AsyncResource):
    """Asynchronous agent-services namespace."""

    async def create(
        self,
        *,
        organization_id: str,
        merchant_id: str,
        product_id: str,
        name: str,
        description: str,
        endpoint: str,
        price: str,
        network: str,
        input_schema: Mapping[str, Any],
    ) -> Any:
        """Register an agent service. ``POST /v1/agent-services``."""
        return await self._request(
            "POST",
            "/v1/agent-services",
            body=_create_body(
                organization_id=organization_id,
                merchant_id=merchant_id,
                product_id=product_id,
                name=name,
                description=description,
                endpoint=endpoint,
                price=price,
                network=network,
                input_schema=input_schema,
            ),
        )

    async def list(self) -> Any:
        """List agent services. ``GET /v1/agent-services``."""
        return await self._request("GET", "/v1/agent-services")

    async def publish(self, service_id: str) -> Any:
        """Publish an agent service. ``POST /v1/agent-services/:id/publish``."""
        return await self._request(
            "POST", f"/v1/agent-services/{service_id}/publish"
        )

    async def metadata(self, service_id: str) -> Any:
        """Fetch machine-readable metadata. ``GET /v1/agent-services/:id/metadata.json``."""
        return await self._request(
            "GET", f"/v1/agent-services/{service_id}/metadata.json"
        )
